packages/idyll-components/src/scroller.js

```diff
@@ -0,0 +1,171 @@
+  }
+
+  handleStepEnter({ element, index, direction }) {
+    this.SCROLL_STEP_MAP[index] && this.SCROLL_STEP_MAP[index]();
+    let update = { currentStep: index };
+    if (this.SCROLL_NAME_MAP[index]) {
+      update.currentState = this.SCROLL_NAME_MAP[index];
+    }
+    this.props.updateProps && this.props.updateProps(update);
+    if (index === Object.keys(this.SCROLL_STEP_MAP).length - 1) {
+      d3.select('body').style('overflow', 'auto');
+    }
+  }
+
+  handleResize() {
+    this.setState({
+      graphicHeight: window.innerHeight + 'px',
+      graphicWidth: window.innerWidth + 'px',
+    });
+  }
+  handleContainerEnter(response) {
+    if (this.props.disableScroll && (!this.props.currentStep || this.props.currentStep < Object.keys(this.SCROLL_STEP_MAP).length - 1)) {
+      d3.select('body').style('overflow', 'hidden');
+    }
+    this.setState({ isFixed: true, isBottom: false });
+  }
+
+  handleContainerExit(response) {
+    this.setState({ isFixed: false, isBottom: response.direction === 'down'});
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (this.props.currentStep !== nextProps.currentStep) {
+      d3.selectAll(`#idyll-scroll-${this.id} .idyll-step`)
+        .filter(function (d, i) { return i === nextProps.currentStep;})
+        .node()
+        .scrollIntoView({ behavior: 'smooth' });
+    }
+    if (this.props.currentState !== nextProps.currentState) {
+      d3.selectAll(`#idyll-scroll-${this.id} .idyll-step`)
+        .filter(function (d, i) { return nextProps.currentState === this.SCROLL_NAME_MAP[i] })
+        .node()
+        .scrollIntoView({ behavior: 'smooth' });
+    }
+    if (nextProps.disableScroll && (!nextProps.currentStep || nextProps.currentStep < Object.keys(this.SCROLL_STEP_MAP).length - 1)) {
+      d3.select('body').style('overflow', 'hidden');
+    }
+  }
+
+  registerStep(elt, name, val)  {
+    this.SCROLL_STEP_MAP[elt] = val;
+    this.SCROLL_NAME_MAP[elt] = name;
+  }
+
+  render() {
+    const { hasError, updateProps, idyll, children, ...props } = this.props;
+    const { isFixed, isBottom, graphicHeight, graphicWidth } = this.state;
+    return (
+      <div ref={(ref) => this.ref = ref} className="idyll-scroll" id={`idyll-scroll-${this.id}`} style={{position: 'relative'}}>
+        <div className="idyll-scroll-graphic"
+          style={Object.assign({ height: graphicHeight },
+            styles.SCROLL_GRAPHIC,
+            isFixed ? styles.SCROLL_GRAPHIC_FIXED : {},
+            isBottom ? styles.SCROLL_GRAPHIC_BOTTOM : {})} >
+
+          <div style={Object.assign({ width: graphicWidth }, styles.SCROLL_GRAPHIC_INNER)}>
+            {filterChildren(
+              children,
+              (c) => {
+                return c.type.name && c.type.name.toLowerCase() === 'graphic';
+              }
+            )}
+          </div>
+        </div>
+        <TextContainer idyll={idyll}>
+          <div className="idyll-scroll-text">
+            {mapChildren(filterChildren(
```

Comment:
Curious about this pattern (also used a few times in `Stepper`) of dynamically binding these parent methods to props of cloned children. Is this the standard way of approaching this, or could you bind these props more declaratively?

If not, it might make things a bit clearer to add a helper function that maps child name -> function, i.e.:
`mapChildProps({ 'Step': { registerStep: this.registerStep.bind(this) } })`

Reply:
Hmm.. I like that API - something to add to https://github.com/idyll-lang/idyll-component-children?

In particular I'd like to avoid making user's of that module write this garbage:

```
return !c.type.name || c.type.name.toLowerCase() === 'step';
```